### PR TITLE
Add option to save payment during BNMO orders

### PR DESCRIPTION
**Pull request by ds300**
https://artsyproduct.atlassian.net/browse/PURCHASE-819

## What happened here?

1. Moved the 'credit card input + maybe billing address input' parts of the `Payment`/`NewPayment` pages into a single component called `PaymentPicker`.
2. Consolidated all the tests for both the `Payment` and `NewPayment` pages that related to filling in those fields and submitting requests to stripe (to tokenize a credit card) and gravity (to save the credit card token).
3. Mocked out the `PaymentPicker` component for the remaining `Payment` and `NewPayment` tests, for simplicity's sake.
4. Fixed the size of the icon in the CreditCardDetails component (another casualty of the palette icon refactor).
5. Also changed the horizontal spacing in the CreditCardDetails content to match designs.
6. Finally, implemented the option to save credit cards for later use.

## Video

save-credit-card

### With no saved cards

image

image

### With saved cards

image

image

### With saved cards but adding a new one

image

image

## Note
After composing this PR I fixed an issue with the size/alignment of the CreditCardDetails component, so that's fixed. I couldn't be bothered to update all the screenshots and video though.

image

### CODE REVIEW

File: src/Apps/Order/Components/PaymentPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+  }
+
+  getCreditCardId: () => Promise<
+    | { type: "error"; error: string | undefined }
+    | { type: "invalid_form" }
+    | { type: "success"; creditCardId: string }
+  > = async () => {
+    const { creditCardSelection, saveNewCreditCard } = this.state
+    if (creditCardSelection.type === "existing") {
+      return { type: "success", creditCardId: creditCardSelection.id }
```
Comment (sweir27) [inline src/Apps/Order/Components/PaymentPicker.tsx]: This pattern is really nice.

File: src/Apps/Order/Components/PaymentPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+
+  private needsAddress = () => {
+    return this.isPickup() || !this.state.hideBillingAddress
+  }
+}
+
+// Our mess of HOC wrappers is not amenable to ref forwarding, so to expose a
+// ref to the PaymentPicker instance (for getCreditCardId) we'll add an
+// `innerRef` prop which gets sneakily injected here
+const PaymentPickerWithInnerRef: React.SFC<
```
Comment (sweir27) [inline src/Apps/Order/Components/PaymentPicker.tsx]: Wow this is slick and sort of crazy that we have to do this!

File: src/Apps/Order/Components/CreditCardDetails.tsx [hunk 3]
```diff
@@ -7,21 +7,28 @@ export const CreditCardDetails = ({
...
+    />
     <Serif
-      size={["2", "3"]}
+      size={responsive ? ["2", "3"] : "3"}
       color="black100"
       style={{ position: "relative", top: "1px" }}
     >
       •••• {last_digits}
-      &nbsp; Exp {expiration_month}/{expiration_year}
+      &nbsp;&nbsp; Exp {expiration_month.toString().padStart(2, "0")}/
```
[the remark this answers is not in the document]
Comment (sweir27) [reply src/Apps/Order/Components/CreditCardDetails.tsx]: 😮 good catch!

**Comment by sweir27** [comment]
@ds300 This is good to merge! Mind rebasing and then we can go for it?

**Comment by artsyit** [comment]
Deploy preview for *artsy-reaction* ready!

Built with commit 174fe217826595429aa7d2621d0101e3a4498fb2

https://deploy-preview-2275--artsy-reaction.netlify.com

**Comment by artsyit** [comment]
:rocket: PR was released in v15.14.3 :rocket:

**Pull request merged**